File: problems/56/56.merge-intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#


# @lc code=start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # intervals[i] = [start, end]
        # merge all overlapping intervals, and return intervals
        # ie:
        # intervals = [[1,3],[2,6],[8,10],[15,18]]
        # [[1,6],[8,10],[15,18]]
        # sort in order of start time

        n = len(intervals)
        intervals.sort(key=lambda x: x[0])
        res = [intervals[0]]

        for i in range(1, n):
            start, end = intervals[i]
            if start <= res[-1][1]:
                res[-1][1] = max(res[-1][1], end)
            else:
                res.append([start, end])

        return res


# @lc code=end

File: problems/56/56.merge-intervals.go
/*
 * @lc app=leetcode id=56 lang=golang
 *
 * [56] Merge Intervals
 */

package solution

import "sort"

// @lc code=start
func merge(intervals [][]int) [][]int {
	// compare prev end to curr start
	// if curr start <= prev end, prev end = max(prev end, curr end)
	// for case: [[1,5],[2,4]]
	// O(NlogN) time, O(N) space

	// sort in ascending order by start time
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	n := len(intervals)
	res := make([][]int, 1)
	res[0] = intervals[0]

	for i := 1; i < n; i++ {
		curr := intervals[i]
		if curr[0] <= res[len(res)-1][1] {
			res[len(res)-1][1] = max(res[len(res)-1][1], curr[1])
		} else {
			res = append(res, curr)
		}
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end
